### PR TITLE
Changes for opting long url or null on link creation  error

### PR DESCRIPTION
**Pull request by sojanpr**
additional APIs added with `defaultToLongUrl`  for creating links sync / async

@aaustin @EvangelosG

### CODE REVIEW

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java [hunk 1]
```diff
@@ -501,15 +501,56 @@ public void registerView(@Nullable RegisterViewStatusListener callback) {
...
+     * @param linkProperties An object of {@link LinkProperties} specifying the properties of this link
+     * @return A {@link String} with value of the short url created for this BUO. A long url for the BUO is returned in case link creation fails
+     */
     public String getShortUrl(@NonNull Context context, @NonNull LinkProperties linkProperties) {
         return getLinkBuilder(context, linkProperties).getShortUrl();
     }
 
+    /**
+     * Creates a short url for the BUO synchronously.
+     * xa
```
Comment (chandsie) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: looks like an accidental comment

File: Branch-SDK/src/io/branch/referral/BranchShortLinkBuilder.java [hunk 2]
```diff
@@ -127,7 +127,19 @@ public BranchShortLinkBuilder setType(int type) {
...
     public String getShortUrl() {
-        return super.getUrl();
+        return super.getUrl(true);
+    }
+
+    /**
+     * <p>Configures and requests a short URL to be generated by the Branch servers, via a synchronous
+     * call; with a duration specified within which an app session should be matched to the link.</p>
+     *
+     * @param defaultToLongUrl A {@link Boolean} whose value should be false if a  long url is not intended to return on failure.
```
Comment (chandsie) [inline Branch-SDK/src/io/branch/referral/BranchShortLinkBuilder.java]: a bit nitpicky: the argument isn't a Boolean, but a primitive boolean (lowercase b)

Comment (chandsie) [inline Branch-SDK/src/io/branch/referral/BranchShortLinkBuilder.java]: Also nitpicky, there's an extra space between "false if a" and "long url"

File: Branch-SDK/src/io/branch/referral/BranchShortLinkBuilder.java [hunk 3]
```diff
@@ -137,10 +149,23 @@ public String getShortUrl() {
...
-        super.generateUrl(callback, isFromShareSheet);
+    /**
+     * <p>Configures and requests a short URL to be generated by the Branch servers, via a asynchronous
+     * call; The {@link Branch.BranchLinkCreateListener} is called back with the url when the url is generated.</p>
+     *
+     * @param callback         A {@link Branch.BranchLinkCreateListener} callback instance that will trigger
+     * @param defaultToLongUrl A {@link Boolean} whose value should be false if a  long url is not intended to return on failure.
+     *                         Setting this it false will return NULL for url on failure else a long url.
+     */
+    public void generateShortUrl(Branch.BranchLinkCreateListener callback, Boolean defaultToLongUrl) {
```
Comment (chandsie) [inline Branch-SDK/src/io/branch/referral/BranchShortLinkBuilder.java]: I believe the `defaultToLongUrl` argument can be just a primitive boolean here. This overhead is probably negligible but consistency would be nice.

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java [hunk 4]
```diff
@@ -501,15 +501,56 @@ public void registerView(@Nullable RegisterViewStatusListener callback) {
...
+     *
+     * @param context          {@link Context} instance
+     * @param linkProperties   An object of {@link LinkProperties} specifying the properties of this link
+     * @param defaultToLongUrl A {@link Boolean} specifies if a long url should be returned in case of link creation error
```
Comment (chandsie) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: a bit nitpicky: the argument isn't a Boolean, but a primitive boolean (lowercase b).
Mentioned elsewhere as well

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java [hunk 5]
```diff
@@ -501,15 +501,56 @@ public void registerView(@Nullable RegisterViewStatusListener callback) {
...
+     *                         If set to false NULL is returned in case of link creation error
```
Comment (chandsie) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: nitpicky again (sorry!), but need a comma between false and NULL.

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java [hunk 6]
```diff
@@ -501,15 +501,56 @@ public void registerView(@Nullable RegisterViewStatusListener callback) {
...
         getLinkBuilder(context, linkProperties).generateShortUrl(callback);
     }
 
+    /**
+     * Creates a short url for the BUO asynchronously.
+     *
+     * @param context          {@link Context} instance
+     * @param linkProperties   An object of {@link LinkProperties} specifying the properties of this link
+     * @param callback         An instance of {@link io.branch.referral.Branch.BranchLinkCreateListener} to receive the results
+     * @param defaultToLongUrl A {@link Boolean} specifies if a long url should be returned in case of link creation error
```
Comment (chandsie) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: argument is boolean, not Boolean

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java [hunk 7]
```diff
@@ -501,15 +501,56 @@ public void registerView(@Nullable RegisterViewStatusListener callback) {
...
+     *                         If set to false NULL is returned in case of link creation error
```
Comment (chandsie) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: same comma nitpick as before

**Comment by sojanpr** [comment]
@chandsie  Thank you for the great review. Everything in the review comments makes perfect sense and we have added fix for it. We have primitive type vs Object issue in quite a few places in documentation. Addressed only for the files in this PR now but will be taken care for other files soon.

Thank you for your efforts to make Branch SDK better.

**Comment by aaustin** [comment]
👍

**Pull request merged**